src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -2036,17 +2047,45 @@ private void VisitAndPushArguments(ImmutableArray<IArgumentOperation> arguments,
         {
             var previousInterpolatedStringHandlerContext = _currentInterpolatedStringHandlerArgumentContext;
 
-            if (arguments.SelectAsArray(predicate: arg => arg.Value is IInterpolatedStringHandlerCreationOperation,
-                                        selector: arg => (IInterpolatedStringHandlerCreationOperation)arg.Value)
-                    is { IsDefaultOrEmpty: false } interpolatedStrings)
+            ArrayBuilder<IInterpolatedStringHandlerCreationOperation>? interpolatedStringBuilder = null;
+            int lastIndexForSpilling = -1;
+
+            for (int i = 0; i < arguments.Length; i++)
             {
+                if (arguments[i].Value is IInterpolatedStringHandlerCreationOperation creation)
+                {
+                    lastIndexForSpilling = i;
+                    interpolatedStringBuilder ??= ArrayBuilder<IInterpolatedStringHandlerCreationOperation>.GetInstance();
+                    interpolatedStringBuilder.Add(creation);
+                }
+            }
+
+            if (lastIndexForSpilling > -1)
+            {
+                Debug.Assert(interpolatedStringBuilder != null);
                 _currentInterpolatedStringHandlerArgumentContext = new InterpolatedStringHandlerArgumentsContext(
-                    interpolatedStrings,
+                    interpolatedStringBuilder.ToImmutableAndFree(),
                     _evalStack.Count - (instancePushed ? 1 : 0),
                     hasReceiver: instancePushed);
             }
 
-            VisitAndPushArray(arguments, UnwrapArgument);
+            for (int i = 0; i < arguments.Length; i++)
+            {
+                PushOperand(VisitRequired(arguments[i].Value));
+
+                // If there are declaration expressions in the arguments before an interpolated string handler, and that declaration
+                // expression is referenced by the handler constructor, we need to spill it to ensure the declaration doesn't end
+                // up in the tree twice. However, we don't want to generally introduce spilling for these declarations: that could
+                // have unexpected affects on consumers. So we limit the spilling to those indexes before the last interpolated string
+                // handler. We _could_ limit this further by only spilling declaration expressions if the handler in question actually
+                // referenced a specific declaration expression in the argument list, but we think that the complexity in implementing
+                // this check is more complexity than this scenario needs.
+                if (i < lastIndexForSpilling)
+                {
+                    SpillEvalStack(spillDeclarationExpressions: true);
```

Comment:
>SpillEvalStack(spillDeclarationExpressions: true); [](http://example.com/codeflow?start=20&length=50)

Instead of doing this, I would simply conditionally unwrap declaration expressions before calling ```VisitRequired``` above.

Reply:
Leaving the logic in SpillEvalStack feels fragile because it will continue to be a trap for user that might think that passing ```true``` actually provides guarantees on its own. Another shortcoming is that this call is going to unwrap declaration expressions that do not belong to this argument list. For example, arguments of an enclosing call. Please add a test case for this scenario.